### PR TITLE
Added optional audience and resource fields for oauth2 datasources

### LINKED ISSUE #4666: [Feature] Allow client credentials flow to add custom body parameters for Auth0
To enable using the client credentials flow provided by `Auth0`, we need to add the `audience` field to the body. To solve this, we can expose body parameters as a field in the REST datasource form.

Ref: https://auth0.com/docs/flows/call-your-api-using-the-client-credentials-flow

### PR DESCRIPTION
**Pull request by nidhi-nair**
Adds the `audience` and `resource` fields to the token request when client credentials are being shared via body. When shared as a basic token in the authorization header, these fields are not available.

This field is stored as a string, and hence always added to the request as is. This means that the if the provider expects any form of encoding for the field, this should be handled on the Appsmith client itself. Similarly, in case multiple audiences or resources are required, the string expects to have the proper stringified array representation.

Screenshot 2021-05-25 at 8 56 55 PM

Fixes #4666 

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
-Tested against mocklab endpoint

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/oauth2-audience-resource 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 51.55 **(0)** | 32.44 **(-0.02)** | 29.67 **(0)** | 52.06 **(0)**
 :red_circle: | app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx | 28.09 **(-0.48)** | 22.92 **(-0.48)** | 3.23 **(-0.1)** | 31.61 **(-0.42)**</details>

### CODE REVIEW

**Review by sumitsum — DISMISSED** [review]

File: app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx [hunk 1]
```diff
@@ -664,6 +687,8 @@ class DatasourceRestAPIEditor extends React.Component<Props> {
             ]}
           />
         </FormInputContainer>
+        {!_.get(formData.authentication, "isAuthorizationHeader") &&
```
Comment (hetunandu) [inline app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx:690]: Set a default value so you will always get false if this field does not exist for some reason
```suggestion
        {!_.get(formData.authentication, "isAuthorizationHeader", false) &&
```

Comment (nidhi-nair) [reply app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx:690]: I'm defaulting `isAuthorizationHeader` to true when the incoming data is undefined. Should this still be required? @hetunandu

**Comment by jeznag** [comment]
"This means that the input field needs to be a properly encoded URI"

I don't think this is true for audience. You can set any string for audience. (e.g. we have api.dev.okrasolar.com)

image

**Comment by nidhi-nair** [comment]
@jeznag You are right, providers are free to choose what type of input they expect to receive for the `audience` field. I meant to say that if there is any encoding required, it will need to be handled on the client, the execution for this API will use the value as is. I'll update the PR description to reflect this. :)

**Pull request merged**